From a0943bd277ed9b5ecaf1c7c1ace7590b1b6da37d Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH] jni: consider /data/app to the fd whitelist if Xposed is detected

Latest security update has added whitelisting routine to the Zygote.

Since Xposed now reads from /data/app,
it's necessary to add /data/app to the whitelist.

Dynamically do this if XposedBridge.jar is detected.

diff --git a/core/jni/fd_utils.cpp b/core/jni/fd_utils.cpp
index 3c1c0bc..0d9e0e0 100644
--- a/core/jni/fd_utils.cpp
+++ b/core/jni/fd_utils.cpp
@@ -111,6 +111,16 @@ class FileDescriptorInfo {
       && path.find("/../") == std::string::npos) {
     return true;
   }
+
+  if (access("/system/framework/XposedBridge.jar", F_OK ) != -1) {
+    // Xposed-powered Zygote might read from extensions other than .apk
+    // so skip extension check
+    // ALOGW("Xposed detected, loosening up Zygote fd check!");
+    static const char* kDataAppPrefix = "/data/app/";
+    if (android::base::StartsWith(path, kDataAppPrefix)) {
+      return true;
+    }
+  }
 
   return false;
 }
